csrc/dynamic_transform.cpp

```diff
@@ -0,0 +1,510 @@
+                       ->input(0)
+                       ->definition()
+                       ->toString();
+    }
+
+    // Replace the old tensor with the new concretized tensor
+    for (auto use_of_old_tv : incomplete_out_tv->uses()) {
+      std::cerr << "Before replacement: " << use_of_old_tv->toString();
+      auto new_use = ir_utils::replaceValInExpr(
+          use_of_old_tv, incomplete_out_tv, concrete_reshape_out_tv);
+      std::cerr << "After replacement: " << new_use->toString();
+    }
+
+    if (incomplete_out_tv->isFusionOutput()) {
+      incomplete_out_tv->fusion()->removeOutput(incomplete_out_tv);
+      incomplete_out_tv->fusion()->addOutput(concrete_reshape_out_tv);
+      incomplete_out_tv->fusion()->removeVal(incomplete_out_tv);
+    }
+  }
+}
+
+namespace {
+
+bool hasSymbolicAxis(const std::vector<IterDomain*>& ids) {
+  return std::any_of(ids.begin(), ids.end(), [](IterDomain* id) {
+    return id->getIterType() == IterType::Symbolic;
+  });
+}
+
+} // namespace
+
+void DynamicTransformConcretizer::mutate(TensorView* tv) {
+  auto propagated = propagateFromProducerToConsumer(tv);
+
+  if (!propagated) {
+    return;
+  }
+
+  // At this point, there should be no expr beyond rfactor root
+  TORCH_INTERNAL_ASSERT(
+      tv->domain()->domain() == tv->getMaybeRFactorDomain(),
+      "Invalid tensor: ",
+      tv->toString());
+
+  // If it has a rfactor root domain, instead of using the
+  // OptOutMutator interface, replay the exprs with the mutated root
+  // IDs and then register the mutated rf domain IDs. This simplifies
+  // the logic as new IterDomains are automatically generated by the
+  // replay.
+  if (tv->hasRFactor()) {
+    std::unordered_map<IterDomain*, IterDomain*> replacement_map;
+    for (auto root_id : tv->getRootDomain()) {
+      auto mutated = maybeMutated(root_id)->as<IterDomain>();
+      TORCH_INTERNAL_ASSERT(mutated != root_id);
+      replacement_map.emplace(root_id, mutated);
+      std::cerr << "Initial replay map: " << root_id->toString() << " -> "
+                << mutated->toString() << std::endl;
+    }
+    auto replay =
+        ReplayTransformations(tv->getMaybeRFactorDomain(), replacement_map)
+            .setReplayResize(true)
+            .getReplay();
+    std::vector<IterDomain*> updated_rf_domain;
+    for (auto rf_id : tv->getMaybeRFactorDomain()) {
+      auto it = replay.find(rf_id);
+      if (it == replay.end()) {
+        for (auto kv : replay) {
+          std::cerr << "Replay: " << kv.first->toString() << ", "
+                    << kv.second->toString() << std::endl;
+        }
+      }
+      TORCH_INTERNAL_ASSERT(
+          it != replay.end(), "Replay of ", rf_id->toString(), " not found");
+      auto updated_rf_id = replay.at(rf_id);
+      registerMutation(rf_id, updated_rf_id);
+    }
+  }
+
+  // Root and rfactor domains are updated. First mutate the
+  // TensorDomain and then TensorView
+  mutate(tv->domain());
```

Comment:
Seems like it would be nice if opt out mutator also covered transformations between root and domain of tensor domains.